homeassistant/components/niu/sensor.py

```diff
@@ -0,0 +1,160 @@
+}
+
+# Sensors present in dual-battery vehicles
+SENSORS_DUAL = {
+    "level a": ("Battery A Level", PERCENTAGE, "mdi:battery"),
+    "level b": ("Battery B Level", PERCENTAGE, "mdi:battery"),
+    "temp a": ("Battery A Temperature", TEMP_CELSIUS, "mdi:thermometer"),
+    "temp b": ("Battery B Temperature", TEMP_CELSIUS, "mdi:thermometer"),
+}
+
+
+async def async_setup_entry(hass, config, add_entities, discovery_info=None):
+    """Set up the sensor platform."""
+
+    coord = hass.data[DOMAIN][config.entry_id]["coordinator"]
+
+    for serial, vehicle in (
+        hass.data[DOMAIN][config.entry_id]["account"].get_vehicles().items()
+    ):
+        entities = []
+
+        for key, value in SENSORS.items():
+            entities.append(NiuSensor(serial, coord, key, value[0], value[1], value[2]))
+
+        for key, value in (
+            SENSORS_SINGLE.items()
+            if vehicle.battery_count == 1
+            else SENSORS_DUAL.items()
+        ):
+            entities.append(NiuSensor(serial, coord, key, value[0], value[1], value[2]))
+
+        add_entities(entities, True)
+
+
+class NiuSensor(NiuVehicle, Entity):
+    """Representation of a Sensor."""
+
+    def __init__(self, vehicle_id, coordinator, attribute, name, unit, icon):
+        """Initialize the sensor."""
+        super().__init__(vehicle_id, coordinator)
+
+        self._attribute = attribute
+        self._name = name
+        self._unit = unit
+        self._icon = icon
+
+    @property
+    def unique_id(self) -> str:
+        """Return the unique id for the sensor."""
+        return f"{self._vehicle.serial_number}_{self._attribute}"
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return f"{self._vehicle.name} {self._name}"
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+
+        # Battery level
+        if self._attribute == "level":
+            return self._vehicle.soc()
+        if self._attribute == "level a":
+            return self._vehicle.soc(0)
+        if self._attribute == "level b":
+            return self._vehicle.soc(1)
+
+        # Odometer
+        if self._attribute == "odometer":
+            return self._vehicle.odometer
+
+        # Range
+        if self._attribute == "range":
+            return self._vehicle.range
+
+        # Charging time
+        if self._attribute == "charging time":
+            return self._vehicle.charging_time_left
+
+        # Temperature
+        if self._attribute == "temp" or self._attribute == "temp a":
+            return self._vehicle.battery_temp(0)
+        if self._attribute == "temp b":
+            return self._vehicle.battery_temp(1)
+
+        return None
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement."""
+        return self._unit
+
+    @property
+    def icon(self):
+        """Return the icon of the sensor."""
+
+        if "level" in self._attribute:
```

Comment:
You should not add the battery icon logic here, since it is handled by the front-end already. Just make sure you set the device class to `battery`.

https://github.com/home-assistant/frontend/blob/dev/src/common/entity/sensor_icon.ts
https://github.com/home-assistant/frontend/blob/dev/src/common/entity/battery_icon.ts